lib/ex_double_entry/schemas/account_balance.ex

```diff
@@ -105,8 +114,30 @@ defmodule ExDoubleEntry.AccountBalance do
 
   def lock_multi!(accounts, fun) do
     ExDoubleEntry.repo().transaction(fn ->
-      accounts |> Enum.sort() |> Enum.map(fn account -> lock!(account) end)
-      fun.()
+      locked_accounts =
+        accounts
+        |> Enum.sort()
+        |> Enum.reduce(from(record in AccountBalance), fn
+          %Account{scope: nil} = account, query ->
+            Ecto.Query.or_where(query, [
+              {:currency, ^account.currency},
+              {:identifier, ^account.identifier}
+            ])
+
+          %Account{} = account, query ->
+            Ecto.Query.or_where(query, [
+              {:currency, ^account.currency},
+              {:identifier, ^account.identifier},
+              {:scope, ^account.scope}
+            ])
+        end)
+        |> lock_cond(true)
+        |> ExDoubleEntry.repo().all()
+
+      accounts
+      |> Stream.map(fn account -> Enum.find(locked_accounts, &(&1.id == account.id)) end)
+      |> Enum.map(&Account.present/1)
+      |> fun.()
```

Comment:
@smt116 could you please help me understand what these changes are? (I'm referring to the entire block of logic.) 🤔

Reply:
Essentially, there are two changes here.

The first one is locking all accounts in a single query instead of doing it each by each. This change is not so important from the concurrency point of view (I was wrong when typing a commit message). Although it uses a single SQL query instead of N which is a small performance improvement.

The second one is making that previous query return accounts and pass them into the callback. This is important from consistency point of view. Previously, the transfer function was re-using the given accounts (as arguments) to calculate balances (e.g., for line items). The problem is that these could be already outdated (modified by another transfer). The idea here is to use the up-to-date data fetched within the same database transaction (used for locking rows) instead of "reloading" records within the `transfer` function.

Reply:
Makes sense 👍